src/eigsolve.jl

```diff
@@ -11,19 +11,82 @@ else
+@doc raw"""
+    struct EigsolveResult{T1<:Vector{<:Number}, T2<:AbstractMatrix{<:Number}, ObjType<:Union{Nothing,OperatorQuantumObject,SuperOperatorQuantumObject}}
+        values::T1
+        vectors::T2
+        type::ObjType
+        dims::Vector{Int}
+        iter::Int
+        numops::Int
+        converged::Bool
+    end
+
+A struct containing the eigenvalues, the eigenvectors, and some information from the solver
+
+# Fields
+- `values::AbstractVector`: the eigenvalues
+- `vectors::AbstractMatrix`: the transformation matrix (eigenvectors)
+- `type::Union{Nothing,QuantumObjectType}`: the type of [`QuantumObject`](@ref), or `nothing` means solving eigen equation for general matrix
+- `dims::Vector{Int}`: the `dims` of [`QuantumObject`](@ref)
+- `iter::Int`: the number of iteration during the solving process
+- `numops::Int` : number of times the linear map was applied in krylov methods
+- `converged::Bool`: Whether the result is converged
+
+# Examples
+One can obtain the eigenvalues and the corresponding [`QuantumObject`](@ref)-type eigenvectors by:
+```
+julia> result = eigenstates(sigmax());
 
-struct EigsolveResult{T1<:Vector{<:Number}, T2<:AbstractMatrix{<:Number}}
-    vals::T1
-    vecs::T2
+julia> λ, ψ, T = result;
+
+julia> λ
+2-element Vector{ComplexF64}:
+ -1.0 + 0.0im
+  1.0 + 0.0im
+
+julia> ψ
+2-element Vector{QuantumObject{Vector{ComplexF64}, KetQuantumObject}}:
+ QuantumObject{Vector{ComplexF64}, KetQuantumObject}(ComplexF64[-0.7071067811865475 + 0.0im, 0.7071067811865475 + 0.0im], KetQuantumObject(), [2])
+ QuantumObject{Vector{ComplexF64}, KetQuantumObject}(ComplexF64[0.7071067811865475 + 0.0im, 0.7071067811865475 + 0.0im], KetQuantumObject(), [2])
+
+julia> T
+2×2 Matrix{ComplexF64}:
+ -0.707107+0.0im  0.707107+0.0im
+  0.707107+0.0im  0.707107+0.0im
+ ```
+"""
+struct EigsolveResult{T1<:Vector{<:Number}, T2<:AbstractMatrix{<:Number}, ObjType<:Union{Nothing,OperatorQuantumObject,SuperOperatorQuantumObject}}
+    values::T1
+    vectors::T2
+    type::ObjType
+    dims::Vector{Int}
     iter::Int
     numops::Int
     converged::Bool
 end
 
-Base.iterate(res::EigsolveResult) = (res.vals, Val(:vecs))
-Base.iterate(res::EigsolveResult, ::Val{:vecs}) = (res.vecs, Val(:done))
+Base.iterate(res::EigsolveResult) = (res.values, Val(:vector_list))
+Base.iterate(res::EigsolveResult{T1,T2,Nothing}, ::Val{:vector_list}) where {T1, T2} = ([res.vectors[:, k] for k in 1:length(res.values)], Val(:vectors))
+function Base.iterate(res::EigsolveResult{T1,T2,ObjType}, ::Val{:vector_list}) where {T1,T2,ObjType<:Union{OperatorQuantumObject,SuperOperatorQuantumObject}} 
+    if res.type isa OperatorQuantumObject
+        return ([QuantumObject(res.vectors[:, k], Ket, res.dims) for k in 1:length(res.values)], Val(:vectors))
+    else
+        return ([QuantumObject(res.vectors[:, k], OperatorKet, res.dims) for k in 1:length(res.values)], Val(:vectors))
+    end
+end
+Base.iterate(res::EigsolveResult, ::Val{:vectors}) = (res.vectors, Val(:done))
```

Comment:
I was thinking of returning `U` as a `Qobj` in the beginning
but I thought that for sparse solvers:  U is usually a non-squared matrix.
if we do it in this way, it will make the return object type unstable (depend on `sparse=true` or `false`).

Reply:
Yes, we can leave it like this